### DISCUSSION TITLE
Inversify Discussion

### DISCUSSION THREAD

**Discussion by PodaruDragos** [mscharley/dot#163] [General]
Hey, I arrived here from the chat on `inversify` discord. I am a maintainer with full rights to `inversify` on the github part. 
You stated that you wanted to improve `inversify` so I started this discussion to see about that ( since you left the server, but i did replied to your threads )

What were your plans and what is stopping you to contribute features to `inversify` itself ? 

Now let's go over some general things/ideas:  
1. It was always clear that typescript "native" decorators won't have support for decorator-metadata, or at least not in the sense that we were familiar with the `emitDecoratorMetadata flag`.
2. At this point in time is not possible to achieve compatibility with none of the IOC projects out there since most if not all are using parameter decorators. It is proposed and it's on stage 1 ( maybe it will happen/maybe not ).
3. As per your code implementation I generally like the idea and would really like to have some sort of experimental "native" container for the tc39 decorators.
4. Combining typescript 5.2's decorator metadata with the approach you did here might be a more "native" way to go forward, even if it's here or in `inversify`. 

Would you like to add some of the things you learned here to an experimental container in `inversify` ?
Would you like to just improve `inversify` as it is ?

If so reach out to me, for contributing permissions.

PS: Thanks for considering and for reading this.

**Comment by mscharley** [comment]
Hi Podaru, thanks for reaching out!

Firstly, let me start by saying that I do like `inversify` and it's served me well for years and I appreciate all the time the maintainers have put into it. I will also admit up front that I haven't looked into the details of how inversify is implemented, I have only ever used it as a user and the public API is the only thing I know about it. Some of the rest of this post may be slightly ignorant as a result, apologies in advance.

That said, lets be direct and address the elephant in the room first:

> Hey, I arrived here from the chat on inversify discord. I am a maintainer with full rights to inversify on the github part.
You stated that you wanted to improve inversify so I started this discussion to see about that ( since you left the server, but i did replied to your threads )
>
> What were your plans and what is stopping you to contribute features to inversify itself ?

When I joined the inversify Discord, you seemed pretty clear that at that time you were a maintainer but that you didn't have merge rights on Github, nor release access in NPM. I was there for a week or two, and there wasn't a response forthcoming from anyone who did have that kind of access so I eventually left again after leaving contact details if the situation changed.

It hasn't gone unnoticed that the situation has changed, but for better or worse we'd already moved on, both for my personal projects and at my workplace. The irony of the situation isn't lost on me - we started rolling out migrations to `dot` the same week that `inversify` finally released version 6.0.2 with inversify/InversifyJS#1499 merged which we'd been waiting for since at least May.

There isn't any bad blood there, I completely understand that sometimes projects go belly up and that's ok and that's what this project was a response to. I'm personally glad to see inversify starting to spin up and see development again!

That said, the PR mentioned above changed one line, was submitted in February and the version of TypeScript that broke without the fix released publicly in March. The PR was approved once in April, but wasn't finally merged until August and didn't get a release until October. That's not a timeline we can deal with for a change that completely breaks our ability to use something as fundamental our IOC container. There's a healthy level of caution around inversify from us at the moment.

Personally, I'm cautiously optimistic about the project, but time will tell. Any insight you can provide around what happened would be appreciated, there should be an email address on my Github profile if you'd rather reach out in private to discuss that.

> 3. As per your code implementation I generally like the idea and would really like to have some sort of experimental "native" container for the tc39 decorators.

Thanks! I quite like it, but I'm biased. It's always good to hear compliments like this. I'm not sure exactly what you're proposing here, but if it's a second `Container` class in `inversify` that's labelled experimental somehow, then personally I think that everyone is served better by just having two libraries. I'm not against a more official affiliation, but instead of having two different types of container in the one library I feel like it'd be better to leave them separate. People can use the newer 'experimental' approach from this library if they want to, or use the original `inversify` container, or install both and use both if they have some kind of use-case where that would make sense. I haven't actually tried it, but the dot `Token` should actually be compatible with the tokens I added via the `inversify-token` plugin.

> 2. At this point in time is not possible to achieve compatibility with none of the IOC projects out there since most if not all are using parameter decorators. It is proposed and it's on stage 1 ( maybe it will happen/maybe not ).

I went into this project knowing that if I was starting a new project then I wanted to support the newer TC39 standard as well as the older experimental TypeScript decorators. That meant making some breaking changes compared the `inversify` API and I leaned into that. There are quite a few breaking changes scattered around the place that mean `dot` isn't a drop in replacement, rather it is designed closer to a spiritual successor. Inspired by, but not beholden to the API design of `inversify`. I don't think I've fully documented anywhere the exact breaking changes but off the top of my head:

* `Container.get()` is always async and any binding may be async - the container works out the details in-between.
* `bind().toDynamicValue()` takes two parameters and must declare what dependencies it needs explicitly.
* `bind().toFactory()` also has the above point, but is almost certainly completely different in concept to the function of the same name in `inversify` as I never fully understood what it was designed for in `inversify`.

I tried to keep things as similar as possible, partly selfishly to reduce my own workload migrating things, but I also didn't shy away from making breaking changes and (in my opinion) I think there's some nice benefits to the updates, for example:

* `Container.validate()` will statically check that your container actually has a valid set of dependencies bound and that you haven't forgotten to bind something. You can run this at application boot time, or even potentially as part of CI if you construct a container and then validate it. This adds a layer of safety sorely missing in `inversify` currently.

If constructor parameter decorators get fully added to the TC39 spec then in principle I'm not against adding them here as an option, but we'd have to see the spec and assess how feasible it would actually be whenever that happens. I'm building a library to release now though, so I have to use what's available now.

> 4. Combining typescript 5.2's decorator metadata with the approach you did here might be a more "native" way to go forward, even if it's here or in inversify.

Thanks for the heads up! I missed this in the 5.2 release notes. For now, it's locked behind a lib (`"esnext.decorators"`) and if downstream applications need that lib defined to use this library then I'm hesitant to use it right now given that there's a perfectly usable implementation that's already working without that extra requirement already. Keep in mind that an explicit goal of this project currently is to support *both* TC39 decorators and experimental decorators for the foreseeable future.

> Would you like to add some of the things you learned here to an experimental container in inversify? Would you like to just improve inversify as it is? What were your plans and what is stopping you to contribute features to inversify itself?

My biggest problem is that these are actually pretty divergent projects at this point and we'd already migrated over before it was obvious that inversify wasn't abandonware anymore. I'm definitely interested in at least talking with the inversify team more and I'll probably rejoin the discord in the coming days if that's an active communication method now.

**Comment by PodaruDragos** [comment]
hey @mscharley thanks for the great response. 

I'll start with the maintenance issues. As of now we are currently 3 members who have full access on github part. So when we do need to merge things or re-trigger the CI's we can, however for example I still don't have the rights to npm ( neither I want them to be fair ). At the time of our previous communication, I did not have fully access so I could have not run or re-trigger CI builds. In regards to communication with the `inversify` team ( which are just some bunch of people *mostly 3 ) our discord channel is clearly the way to go. I do check it out from time to time and try to at least respond to PR comments.

Now for a clarification part of why I started this discussion. Mainly after looking over the code here it was clear help from someone who researched and build a "successor" would be extremely beneficial. I know you even offered help from you and your team ( but at the time I had no way of actually doing much in terms of PR's ) that's why I opened this discussion after I sorted my rights.

The divergence part is very obvious, but some takes from this can also be applied there. Now we can always introduce breaking changes to inversify itself, for example the only way to resolve dependencies will be to add an array of services ( like you did here ).
This can also work with the experimental decorators and should work just fine with the "native" too.

Since you are doing such a good job here, I won't have the audacity to suggest you help us out, but if you ever reconsider feel free to drop me a message on discord.

PS: my apologies for the late response, I was off for a couple of weeks.
